### PR TITLE
Fix stale grid cache reads caused by uninitialized blocks.

### PR DESCRIPTION
**Pull request by jamii**
Contains an immediate fix for https://github.com/tigerbeetledb/tigerbeetle/issues/345, along with many additional asserts which helped track down the bug.

### CODE REVIEW

**Comment by jamii** [comment]
Currently blows up on https://github.com/tigerbeetledb/tigerbeetle/pull/346/files#diff-e393c4fc086967a9436875ec0217d7f613603e65c0653405e8da5ec824e46dd7R303.

```
>>> p k
$1 = {
  id = 1800825,
  static sentinel_key = {
    id = 18446744073709551615,
    static sentinel_key = <same as static member of an already seen type>
  }
}
>>> p key
$2 = {
  id = 1395874,
  static sentinel_key = {
    id = 18446744073709551615,
    static sentinel_key = <same as static member of an already seen type>
  }
}
```

**Comment by jamii** [comment]
Still blowing up in the same spot, despite now completely validating the contents of the manifest down to individual values. So prime suspect is a bug in table_iterator.

**Comment by jamii** [comment]
In Grid.start_read we [reserve a cache entry](https://github.com/tigerbeetledb/tigerbeetle/blob/089776263b6efcc6c66b9ccb437b4eb509ad7ded//src/lsm/grid.zig#L452) but don't write to the block yet - leaving any previously cached value intact. 

The sequence of events in this crash was:

Write address1, block1 to way1.
Remove address1 from way1.
Write address1, block2 to way2.
Write address2 to way1 without setting block. We now have a valid entry for address1, block1 in way1.
Write address1, block3 to way3. This should remove address1, block2 from way2 but actually removes address1, block1 from way1, since that's the first valid entry it finds.

We now have two entries for address1. Havoc ensues.

**Comment by jamii** [comment]
The fix currently in this PR is not correct, just sufficiently less wrong to make this fuzz seed pass and prove that this is the cause.

**Comment by jamii** [comment]
Probably the correct fix is to separate assigning a cache entry from marking it as valid. So the flow for read would be:

Reserve a block from the cache
Storage writes to that block
Mark the block as valid

**Comment by jamii** [comment]
After further consideration, I don't think it's possible to fix this nicely without changing the cache interface. I've proposed that in https://github.com/tigerbeetledb/tigerbeetle/issues/353.

In the meantime, this pr contains an ugly fix that prevents this particular bug but does not remove the underlying footgun.

File: src/lsm/tree.zig [hunk 1]
```diff
@@ -580,6 +580,10 @@ pub fn TreeType(comptime TreeTable: type, comptime Storage: type, comptime tree_
             assert(tree.compaction_io_pending == 0);
             assert(tree.compaction_callback == null);
 
+            if (constants.verify) {
```
Comment (jamii) [inline src/lsm/tree.zig:596]: This check, and the one below, may slow the fuzzer down too much to be worthwhile.

**Review by sentientwaffle — APPROVED** [review]

**Pull request merged**